File: oop/java/src/Lista01/Ex09/Ex09.java
package Lista01.Ex09;

import Lista01.KeyboardInput;

public class Ex09 {
    static final Double ERROR = 10e-7;
    private Integer iterator = 0;

    private Double calculateBaseEquation(Double x) {
        return Math.pow(x, 3) - Math.pow(x, 2) - 13*x + 8;
    }

    private Double deriveBaseEquation(Double x) {
        return 3 * Math.pow(x, 2) - 2 * x - 13;
    }

    public Integer getLastOperationTotalIterations() { return this.iterator; }

    public Double solveByNewtonRaphson(Double attempt) {
        Double previousAttempt;
        do {
            previousAttempt = attempt;
            attempt -= calculateBaseEquation(attempt) / deriveBaseEquation(attempt);
            this.iterator++;
        } while (previousAttempt - attempt >= ERROR);

        return attempt;
    }

    public static void main(String[] args) {
        Ex09 newtonRaphsonCalculator = new Ex09();
        KeyboardInput keyboardInput = new KeyboardInput(System.in);

        System.out.println("Provide the desired attempt: ");
        Double input = keyboardInput.readDouble();

        System.out.printf("Solution by Newton-Raphson: %f\n", newtonRaphsonCalculator.solveByNewtonRaphson(input));
        System.out.printf("%d total iterations", newtonRaphsonCalculator.getLastOperationTotalIterations());
    }
}

File: oop/lista01_py/Ex09.py
from KeyboardInput import KeyboardInput

class Ex09:
    def __init__(self) -> None:
        self.ERROR = 10e-7
        self.__iterator = 0

    def calculateBaseEquation(self, x): return x**3 - x**2 - 13*x + 8
    def deriveBaseEquation(self, x): return 3 * x**2 - 2 * x - 13

    def getLastOperationTotalIterations(self): return self.__iterator

    def solveByNewtonRaphson(self, attempt):
        previousAttempt = attempt
        attempt -= self.calculateBaseEquation(attempt) / self.deriveBaseEquation(attempt)
        self.__iterator = 1

        while previousAttempt - attempt >= self.ERROR:
            previousAttempt = attempt
            attempt -= self.calculateBaseEquation(attempt) / self.deriveBaseEquation(attempt)
            self.__iterator += 1
        return attempt

if __name__ == "__main__":
    newtonRaphsonCalculator = Ex09()
    keyboardInput = KeyboardInput()

    print("Provide the desired attempt: ")
    input = keyboardInput.readFloat()

    print("Solution by Newton-Raphson: {0}\n".format(newtonRaphsonCalculator.solveByNewtonRaphson(input)))
    print("{0} total iterations".format(newtonRaphsonCalculator.getLastOperationTotalIterations()))